### PR TITLE
vscode: support icons contribution point

### PR DESCRIPTION
**Pull request by vladarama**
#### What it does

Fixes: #12355 

This PR adds support for the `contributes.icon` contribution point from VSCode. This fixes a number of issues where extension icons were invisible such as many of the Gitlens icons.

#### How to test

You can test the functionality of this PR by downloading an extension which contributes its icons through the icons contribution point. One such example is Gitlens.

1.  Open Theia application.
2.  Go to the `Extensions` view
3.  Search and download Gitlens from the vsx-registry.
4.  Verify that all contributed icons are visible and correct.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/monaco/src/browser/monaco-icon-registry-types.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+// *****************************************************************************
+
+import { Event } from '@theia/core';
+import { IJSONSchema } from '@theia/core/lib/common/json-schema';
+import { ThemeIcon } from '@theia/monaco-editor-core/esm/vs/platform/theme/common/themeService';
+import { URI } from '@theia/core/shared/vscode-uri';
+
+//  ------ API types
+
+export type IconIdentifier = string;
```
Comment (tsmaeder) [inline packages/monaco/src/browser/monaco-icon-registry-types.ts:24]: Redefining the string type does not really add any type safety. Typescript is not Ada.

File: packages/monaco/src/browser/monaco-icon-registry-types.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+// icon registry
+export const Extensions = {
```
Comment (tsmaeder) [inline packages/monaco/src/browser/monaco-icon-registry-types.ts:27]: This seems unused.

File: packages/monaco/src/browser/monaco-icon-registry-types.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    IconContribution: 'base.contributions.icons'
+};
+
+export type IconDefaults = ThemeIcon | IconDefinition;
```
Comment (tsmaeder) [inline packages/monaco/src/browser/monaco-icon-registry-types.ts:31]: This seems to be the type that happens to be used where a default value for an icon is needed. For this, the name is misleading. The type is used exactly once and it does not describe anything, so I would just remove it.

File: packages/monaco/src/browser/monaco-icon-registry-types.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+    /**
+     * Deregister a icon from the registry.
+     */
+    deregisterIcon(id: IconIdentifier): void;
+
+    /**
+     * Get all icon contributions
+     */
+    getIcons(): IconContribution[];
```
Comment (tsmaeder) [inline packages/monaco/src/browser/monaco-icon-registry-types.ts:82]: this seem unused

File: packages/monaco/src/browser/monaco-icon-registry-types.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+    /**
+     * Get the icon for the given id
+     */
+    getIcon(id: IconIdentifier): IconContribution | undefined;
```
Comment (tsmaeder) [inline packages/monaco/src/browser/monaco-icon-registry-types.ts:87]: Unused

File: packages/monaco/src/browser/monaco-icon-registry-types.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+    /**
+     * JSON schema for an object to assign icon values to one of the icon contributions.
+     */
+    getIconSchema(): IJSONSchema;
```
Comment (tsmaeder) [inline packages/monaco/src/browser/monaco-icon-registry-types.ts:92]: unused

File: packages/monaco/src/browser/monaco-icon-registry-types.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+    /**
+     * JSON schema to for a reference to a icon contribution.
+     */
+    getIconReferenceSchema(): IJSONSchema;
```
Comment (tsmaeder) [inline packages/monaco/src/browser/monaco-icon-registry-types.ts:97]: unused

File: packages/plugin-ext/src/main/browser/plugin-icon-service.ts [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+import { IThemeService } from '@theia/monaco-editor-core/esm/vs/platform/theme/common/themeService';
+import { UnthemedProductIconTheme } from '@theia/monaco-editor-core/esm/vs/platform/theme/browser/iconsStyleSheet';
+
+@injectable()
+export class PluginIconService implements Disposable {
+
+    @inject(IconRegistry)
+    protected readonly iconRegistry: IconRegistry;
+
+    @inject(IconStyleSheetService)
```
Comment (tsmaeder) [inline packages/plugin-ext/src/main/browser/plugin-icon-service.ts:35]: This instnace variable is unused.

File: packages/monaco/src/browser/monaco-icon-style-sheet.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { injectable } from '@theia/core/shared/inversify';
+import { getIconsStyleSheet } from '@theia/monaco-editor-core/esm/vs/platform/theme/browser/iconsStyleSheet';
+import { IconsStyleSheet } from './monaco-icon-registry-types';
+
+@injectable()
+export class MonacoIconStyleSheetService {
```
Comment (tsmaeder) [inline packages/monaco/src/browser/monaco-icon-style-sheet.ts:22]: I don't see any use for this type except in the unused variable in `plugin-icon-service.ts`

File: packages/monaco/src/browser/monaco-icon-registry-types.ts [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    deregisterIconFont(id: string): void;
+
+    /**
+     * Get the icon font for the given id
+     */
+    getIconFont(id: string): IconFontDefinition | undefined;
+}
+
+export const IconsStyleSheet = Symbol('IconsStyleSheet');
```
Comment (tsmaeder) [inline packages/monaco/src/browser/monaco-icon-registry-types.ts:117]: Anything below here seems unused.

**Review by tsmaeder — CHANGES_REQUESTED** [review]
Hi Vlad, thanks for the PR. There seems to be a bit of unused code and the layering (core vs. monaco) needs fixing. Functionally, it seems to do the job.

**Comment by JonasHelming** [comment]
@vladarama Can you tell when you want to look at Thomas comments? I think this PR would be great to have for the upcoming release...

**Comment by vladarama** [comment]
Hey @JonasHelming,
I have been pretty busy with school lately, but I plan on taking a look at it this week.

**Comment by JonasHelming** [comment]
@vladarama Sounds great! Please keep in mind that we have a release on Thursday this week, maybe this one can make it in :-)

**Comment by tsmaeder** [comment]
@vladarama could you address the remaining open comment, please?

File: packages/monaco/src/browser/monaco-icon-registry-types.ts [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    readonly src: IconFontSource[];
+}
+
+export interface IconFontSource {
+    readonly location: URI;
+    readonly format: string;
+}
+
+export const IconRegistry = Symbol('IconRegistry');
+export interface IconRegistry {
```
[the remark this answers is not in the document]
Comment (vladarama) [reply packages/monaco/src/browser/monaco-icon-registry-types.ts]: @tsmaeder 
I just pushed my latest changes. Is that the implementation you had in mind ?

File: packages/monaco/src/browser/monaco-icon-registry.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+import { injectable } from '@theia/core/shared/inversify';
+import { Event } from '@theia/core';
+import { ThemeIcon } from '@theia/monaco-editor-core/esm/vs/platform/theme/common/themeService';
+import { IconRegistry } from '@theia/core/lib/browser/icon-registry';
+import { IconDefinition, IconFontDefinition, getIconRegistry } from '@theia/monaco-editor-core/esm/vs/platform/theme/common/iconRegistry';
+
+@injectable()
+export class MonacoIconRegistry implements IconRegistry {
+
+    readonly onDidChange: Event<void>;
```
Comment (tsmaeder) [inline packages/monaco/src/browser/monaco-icon-registry.ts:26]: Is the "onDidChange" event connected to anything? If not, why do we have it?

File: packages/monaco/src/browser/monaco-icon-registry.ts [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+import { injectable } from '@theia/core/shared/inversify';
+import * as monaco from '@theia/monaco-editor-core/esm/vs/platform/theme/common/iconRegistry';
+import { IconContribution, IconDefaults, IconFontDefinition, IconRegistry } from './monaco-icon-registry-types';
+import { Event } from '@theia/core';
+import { IJSONSchema } from '@theia/core/lib/common/json-schema';
+import { ThemeIcon } from '@theia/monaco-editor-core/esm/vs/platform/theme/common/themeService';
+
+@injectable()
+export class MonacoIconRegistry implements IconRegistry {
```
[the remark this answers is not in the document]
Comment (tsmaeder) [reply packages/monaco/src/browser/monaco-icon-registry.ts:24]: Ah, I see, we can't instantiate it.

**Comment by tsmaeder** [comment]
@vladarama one more question I have, but then I think we're good to go.

File: packages/monaco/src/browser/monaco-icon-registry.ts (hunk 12 above)
Comment (vladarama) [reply packages/monaco/src/browser/monaco-icon-registry.ts]: Thanks for the reviews Thomas,
The event doesn't seem to be connected to anything. Its been removed 👍

**Review by tsmaeder — APPROVED** [review]
looks good to me now.

**Pull request merged**